modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryReaderExImpl.java

```diff
@@ -1364,10 +1364,16 @@ float readFloat(int fieldId) throws BinaryObjectException {
     @Nullable @Override public Object[] readObjectArray() throws BinaryObjectException {
         switch (checkFlag(OBJ_ARR)) {
             case NORMAL:
-                return BinaryUtils.doReadObjectArray(in, ctx, ldr, this, false, true);
+                if (BinaryArray.useBinaryArrays())
+                    return BinaryUtils.doReadBinaryArray(in, ctx, ldr, this, false, true).deserialize(ldr);
+                else
+                    return BinaryUtils.doReadObjectArray(in, ctx, ldr, this, false, true);
 
             case HANDLE:
-                return readHandleField();
+                if (BinaryArray.useBinaryArrays())
```

Comment:
It's safer to use `instanceof BinaryArray` in case we use `doReadObjectArray` somewhere before.

Reply:
Fixed.